File: Array/medium/cpp/set_matrix_zero.cpp
//
// Created by santosh on 12/20/25.
//

#include<bits/stdc++.h>
using namespace std;

class Solution
{
    public:
    void setMatrixZeroBruteForce(vector<vector<int> > &matrix)
    {
        int m= matrix.size();
        int n = matrix[0].size();
        vector<pair<int,int>> zeros;
        for (int i=0; i<m; i++)
        {
            for (int j=0; j<n; j++)
            {
                if (matrix[i][j]==0)
                {
                    zeros.push_back({i,j});
                }
            }
        }

        for (int i=0 ; i<zeros.size() ; i++)
        {
            int row = zeros[i].first;
            int col = zeros[i].second;

            for (int k=0; k<n; k++)
            {
                matrix[row][k]=0;
            }

            for (int k=0; k<m; k++)
            {
                matrix[k][col]=0;
            }
        }

    }
};

int main()
{
    vector<vector<int>> matrix = {
        {1, 1, 1},
        {1, 0, 1},
        {1, 1, 1}
    };

    Solution s;
    s.setMatrixZeroBruteForce(matrix);

    for (int i=0; i<matrix.size(); i++)
    {
        for (int j=0; j<matrix[0].size(); j++)
        {
            cout<<matrix[i][j]<<" ";
        }
        cout<<endl;
    }
}

File: Array/medium/python/set_matrix_zero.py
# Set Matrix Zeroes (LeetCode 73)
# Created by santosh

class Solution:
    def set_matrix_zero_brute_force(self, matrix: list[list[int]]) -> None:
        """
        Brute force: find zeros first, then mark rows and columns.
        Time: O(m*n*(m+n)), Space: O(k) for zero positions
        """
        m = len(matrix)
        n = len(matrix[0])
        zeros = []

        for i in range(m):
            for j in range(n):
                if matrix[i][j] == 0:
                    zeros.append((i, j))

        for row, col in zeros:
            for k in range(n):
                matrix[row][k] = 0
            for k in range(m):
                matrix[k][col] = 0


if __name__ == "__main__":
    matrix = [
        [1, 1, 1],
        [1, 0, 1],
        [1, 1, 1]
    ]
    s = Solution()
    s.set_matrix_zero_brute_force(matrix)
    for row in matrix:
        print(" ".join(map(str, row)))
